### ISSUE TITLE
katib metrics collector solution

### ISSUE THREAD

**Issue by hougangliu** [kubeflow/katib#685]
- [X] metricsCollector API #697
- [x] Pod level webhook for sidecar
- [ ] Job level webhook for sidecar (traced by #928)
- [x] StdOutCollector metricsCollector executor
- [x]	FileCollector metricsCollector executor
- [x] TfEventCollector metricsCollector executor
- [ ] PrometheusMetricCollector metricsCollector executor  **P2**  (traced by #929)

**Comment by hougangliu** [comment]
## Overview
This solution key ideas are:
1. Inject collector container as a sidecar into `Job`/`Tfjob`/`PytorchJob` Pod. Collector container collects metrics from worker container by parsing file or http based on metrics source, and persist metrics into persistent server, such as `katib-db` by `Katib-manager` or `metadata server`;
2. Introduce `TrialMetrics` CRD to trace metrics of trial by Kubernetes native way;
3. Introduce `Katib-Metrics-Manager` server to poll metrics of running Trial from metric persistent server and update related `TrialMetrics` CR. `Trial` controller updates trial's status based on coresponding `TrialMetrics` CR;

As for Early-Stop service (Maybe discuss it in other issue)
1. Early-Stop service will watch `Trial` metrics observation, once meeting goal, marks corresponding experiments as `Completed`.
1. Experiment controller will mark all running `Trial` as `Killed` or else if experiment `Completed`
1. Trial controller stops Job/TFJob/PytorchJob if `Trial` status is `Killed` or else 

See the architecture as below:

<img width="772" alt="image" src="https://user-images.githubusercontent.com/10126721/61432863-4c185e80-a964-11e9-90da-f111316c0751.png">

## Data Structure
```
type ExperimentSpec struct {
	...
	MetricsCollectorSpec *MetricsCollectorSpec `json:"metricsCollectorSpec,omitempty"`
    ...
}

type MetricsCollectorSpec struct {
	Source    *SourceSpec    `json:"source,omitempty"`
	Collector *CollectorSpec `json:"collector,omitempty"`
}

type SourceSpec struct {
    // maybe model-train source code exposes metrics by http, such as HTTP endpoint in prometheus metric format
    HttpGet        *v1.HTTPGetAction `json:"httpGet,omitempty"`
    // during training model, metrics maybe be persistent into local file in source code, such as tfevent
    FileSystemPath *FileSystemPath   `json:"fileSystemPath,omitempty"`
}

type FileSystemKind string
const (
	DirectoryKind          FileSystemKind = "diretory"
	FileKind               FileSystemKind = "file"
)
type FileSystemPath struct {
    Path  string         `json:"path,omitempty"`
    Kind  FileSystemKind `json:"kind,omitempty"`
}

type CollectorKind string
const (
	StdOutCollector           CollectorKind = "stdOutCollector"
	FileCollector             CollectorKind = "fileCollector"
	TfEventCollector          CollectorKind = "tfEventCollector"
	PrometheusMetricCollector CollectorKind = "PrometheusMetricCollector"
	CustomCollector           CollectorKind = "customCollector"
	NoneCollector             CollectorKind = "noneCollector"
)

type CollectorSpec struct {
    Kind            CollectorKind `json:"kind"`
    CustomCollector *v1.Container `json:"customCollector,omitempty"`
}

type TrialMetrics struct {
    metav1.TypeMeta           `json:",inline"`
	metav1.ObjectMeta         `json:"metadata,omitempty"`
	Spec   TrialMetricsSpec   `json:"spec,omitempty"`
}

type TrialMetricsSpec struct {
    Observation *common.Observation `json:"observation,omitempty"`
}
```

1. For collector kind `StdOutCollector`, `FileCollector`, `TfEventCollector` and `PrometheusMetricCollector` (low priority), katib will maintain their collector implementations and build related images;`StdOutCollector`, `FileCollector` will be similar as current default metrics collector, and `TfEventCollector` keeps same as current tfevent metrics collector
2. katib will install a configmap `katib-config` to match images for each collector kind (not including customCollector), experiment controller will inject collecotr sidecar container into TrialTemplate (maybe it's better trial controller injects collecotr sidecar container into woker job)
3. Share metrics file/dir between worker container and collector container by [emptyDir](https://kubernetes.io/docs/concepts/storage/volumes/#emptydir)

## Example
1. For train source code prints metric in stdout, experiment should be like below (maybe if `metricsCollectorSpec` is nil, we can take it as this case by default):
```
apiVersion: "kubeflow.org/v1alpha2"
kind: Experiment
metadata:
  namespace: kubeflow
  name: random-experiment
spec:
  ...
  metricsCollectorSpec:
    collector:
      kind: stdOutCollector
  ...
```
2. For train source code prints metric into a specifed file with katib default format, experiment should be like below:
```
apiVersion: "kubeflow.org/v1alpha2"
kind: Experiment
metadata:
  namespace: kubeflow
  name: random-experiment
spec:
  ...
  metricsCollectorSpec:
    source:
      fileSystemPath:
        kind: file
        path: "/var/log/train.log"
    collector:
      kind: fileCollector
  ...
```
3. For train source code prints metric into a specifed file without katib default format, experiment should be like below (need user implements his custom collector):
```
apiVersion: "kubeflow.org/v1alpha2"
kind: Experiment
metadata:
  namespace: kubeflow
  name: random-experiment
spec:
  ...
  metricsCollectorSpec:
    source:
      fileSystemPath:
        kind: file
        path: "/var/log/train.log"
    collector:
      kind: customCollector
      customCollector:
        image: xx-repo/xx-user/user-custom-collector-image
        ...
  ...
```
4. For tensorflow source code persists metrics by tfevent:
```
apiVersion: "kubeflow.org/v1alpha2"
kind: Experiment
metadata:
  namespace: kubeflow
  name: random-experiment
spec:
  ...
  metricsCollectorSpec:
    source:
      fileSystemPath:
        kind: directory
        path: "/var/log/train/"
    collector:
      kind: tfEventCollector
```
5. For train source code persist its metrics into persistent server directory directly:
```
apiVersion: "kubeflow.org/v1alpha2"
kind: Experiment
metadata:
  namespace: kubeflow
  name: random-experiment
spec:
  ...
  metricsCollectorSpec:
    collector:
      kind: noneCollector
  ...
```

**Comment by gaocegege** [comment]
LGTM, generally. Let's discuss it today

**Comment by wuchunghsuan** [comment]
For injecting metrics collector sidecar, @gaocegege and I have another implementation:
The MutatingWebHook of metrics collector can decide which pod to be injected sidecar based on labels,  for example, the label like `kubeflow.org/replica-role: master`. The labels are tagged by job operators (e.x. TFjob/PyTorchjob). The sidecar will be injected only if the webhook recognize this label.

In this new design,

1. It is more general since the webhook treats every job as the same and inject sidecar based on labels. 
2. Slightly modification in the job operator is needed. Operators will tag the labels.

**Comment by johnugeorge** [comment]
Can you explain what component is referred to "MutatingWebHook of metrics collector"?

**Comment by gaocegege** [comment]
@johnugeorge I think we should say `Mutating Webhook Server of Katib`

**Comment by gaocegege** [comment]
Summary: In the worst case, we use a pod level webhook to inject sidecar into pod. In Kubernetes 1.5, the webhook supports objectSelector, then we do not have the performance problem.

**Comment by hougangliu** [comment]
In istio solution, if a namespace has "istio-injection=enabled" label, all pods in the namespace will be injected by pod level webhook and its performance influence is little.

So I think maybe we can handle it by like this:
1. if a namespace has "katib-metrics-collector-injection=enabled" label, we can just inject sidecar container in pod level.
2. if  a namespace has no "katib-metrics-collector-injection=enabled" label, we can just inject sidecar container in job/tfjob/pytorchjob level

**Comment by gaocegege** [comment]
Yeah, I think so. In the current version, it is the best way I think.

**Comment by hougangliu** [comment]
then @wuchunghsuan please update sidecar injection solution in design doc in two level as above. thanks

**Comment by wuchunghsuan** [comment]
OK, I got it.

**Comment by hougangliu** [comment]
Per talk by slack, in this solution, we just focus on metricsCollector:
1. metricsCollector will collect and persist metrics into persistent layer (for now katib-manager)
2. when a worker completed, trial controller will call DB to fill back final metrics key/value into trial status.

So periodic metrics will be discussed in early-stopping topic and `TrialMetrics` also be dropped here.

**Comment by hougangliu** [comment]
- [X] metricsCollector API #697
- [ ] Pod level webhook for sidecar
- [ ] Job level webhook for sidecar
- [ ] StdOutCollector metricsCollector executor
- [ ]	FileCollector metricsCollector executor
- [ ] TfEventCollector metricsCollector executor
- [ ] PrometheusMetricCollector metricsCollector executor  **P2**

**Comment by gaocegege** [comment]
> Pod level webhook for sidecar

Can you take the task @wuchunghsuan

**Comment by gaocegege** [comment]
Can talk to apiserver, but have some problems
- When a sidecar is injected, cannot finish the training
- How to control the retry behaviour
  - In-container
  - Using K8s mechanism

**Comment by hougangliu** [comment]
> Hi folks what are you planning on delivering in 0.7?

@jlewi I open #728 to trace what we plan to deliver in 0.7

**Comment by hougangliu** [comment]
for this solution, we add a metricCollector sidecar container into worker Pod. metricCollector sidecar container need collect the metrics, it also need know that worker container has finished so that it can exit, too. otherwise the Pod will keep `Running` since metricCollector sidecar container is running.
So we should notify metricCollector sidecar container when to exit. maybe we can implement it by below solutions:
1. [**share process namespace**](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/):  I like this solution, and [argo](https://github.com/argoproj/argo/tree/master/workflow/executor/pns) also supports this option to watch other containers in same Pod, but `share process namespace` upgraded to beta version when 1.12, we should also consider that katib runs on k8s version older than 1.12
2. **k8s api to get/watch Pod**: call k8s client api to watch the worker Pod, and get other containers status by the Pod status's containerStatuses filed.
**2.1**: If it is metricsCollector to watch/get Pod, the worker Pod need have `Pod watch/get role`, since the experiment can be created in any namespace, we must create serviceAccount with the role in corresponding namespace(katib-controller need getOrCreateServiceAccount in the namespace and set it as serviceaccount of the Pod while inject metrics sidecar container, if the original Pod already have a serviceaccount with non-default service-account-token, it is not easy to handle this case since only one service-account-token will be applied in the container).
**2.2**: we need a long-running service (maybe katib-controller or a new service) with `watch/get Pod role`, it has a rest/grpc api which accepts argument (metrics-collector-client-side, podNamespacedName), when a worker Pod starts, metrics-collector container call the rest/grpc api with argument like (`$podIP:8000`, `kubeflow/random-experiment-xztf8nph-master-0`), then katib-controller starts a goroutine to watch the `kubeflow/random-experiment-xztf8nph-master-0` Pod, once it finds that any container except metricsCollectors have terminated, it will send a message to `$podIP:8000` (metricsCollector listen to this port), and when metricsCollectors get this message, it exit with 0. For this solution, the long-running service(maybe katib-controller) may have many goroutine to watch Pod, it consumes much resource.
3. **wrap command of containers in the Pod except metricsCollector container**: this solution is applied by Tekton project, while inject  metricsCollector sidecar container, we should also wrap command of other containers as the new command in the Pod, the wrap function is very easy just like as below, metricsCollector will wait for the signal generated by the wrap function
```
exitcode=`run the original command`;
send an end message to 127.0.0.1:8000 or generate a shared file;
exit $exitcode;
```

@richardsliu @johnugeorge @gaocegege IMO, maybe we have to implement solution 3. but I need your suggestion, too.

**Comment by gaocegege** [comment]
Thanks for the research!

> we should also consider that katib runs on k8s version older than 1.12

I am not sure about it.

Personally, I prefer the first option.

**Comment by hougangliu** [comment]
Per talk with team, we decide to choose **share process namespace** solution to solve this problem

**Comment by johnugeorge** [comment]
Option 1 seems to the best option as it has no other dependency and it is specifically meant for the same purpose.

**Comment by jlewi** [comment]
@hougangliu @johnugeorge any update on this?

**Comment by hougangliu** [comment]
> @hougangliu @johnugeorge any update on this?

@jlewi for now, metrics collector can work well by pod level metricsCollector container sidecar injection (also included in 0.7.0 release).
considering potential performance impact, we plan to also support to inject metricsCollector container in Job/TfJob/PytorchJob level in next release (will start to discuss in katib weekly meeting)

**Comment by gaocegege** [comment]
I think we can safely remove 0.7 label and add 1.0 since high priority features are implemented now.

**Comment by jlewi** [comment]
@gaocegege and @hougangliu should we file more fine grained issues for the remaining work and then close this issue?

**Comment by hougangliu** [comment]
@jlewi #929 and #928 trace the remaining two issue.
